File: src/main/java/net/plan99/nodejs/java/NodeJS.java
package net.plan99.nodejs.java;

import net.plan99.nodejs.kotlin.NodeJSAPI;
import org.graalvm.polyglot.Context;
import org.graalvm.polyglot.TypeLiteral;
import org.graalvm.polyglot.Value;
import org.intellij.lang.annotations.Language;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.Arrays;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Executor;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.function.Supplier;
import java.util.jar.JarInputStream;

/**
 * Provides an interface to the NodeJS runtime for Java developers. You can only access the NodeJS world
 * when running on the event loop thread, which means you must use the various runJS methods on this class
 * to get onto the correct thread before using eval.
 */
@SuppressWarnings("WeakerAccess")
public class NodeJS {
    private static class Linkage {
        LinkedBlockingDeque<Runnable> taskQueue;
        Value evalFunction;
        Thread nodeJSThread;
        Context ctx;

        Linkage(LinkedBlockingDeque<Runnable> taskQueue, Value evalFunction) {
            this.taskQueue = taskQueue;
            this.evalFunction = evalFunction;
            this.nodeJSThread = Thread.currentThread();
            this.ctx = Context.getCurrent();
        }
    }

    private volatile static Linkage linkage;

    // Called from the boot.js file as part of NodeJVM startup, do not call.
    @SuppressWarnings("unused")
    @Deprecated
    public static void boot(LinkedBlockingDeque<Runnable> taskQueue,
                            Value evalFunction,
                            String[] args) throws ClassNotFoundException, IOException {
        try {
            boot1(taskQueue, evalFunction, args);
        } catch (Throwable e) {
            e.printStackTrace();
            System.exit(1);
        }
    }

    private static void boot1(LinkedBlockingDeque<Runnable> taskQueue, Value evalFunction, String[] args) throws ClassNotFoundException, IOException {
        assert linkage == null : "Don't call this function directly. Already started!";
        assert evalFunction.canExecute();
        NodeJS.linkage = new Linkage(taskQueue, evalFunction);
        Thread.currentThread().setName("NodeJS main thread");

        if (args.length == 0) {
            System.err.println("You must specify at least a class name, or -jar jarname.jar");
            System.exit(1);
        } else if (!args[0].equals("-jar")) {
            Class<?> entryPoint = Class.forName(args[0]);
            startJavaThread(entryPoint, Arrays.copyOfRange(args, 1, args.length));
        } else {
            File myJar = new File(args[1]);
            final URL url = myJar.toURI().toURL();
            String mainClassName = "";
            try (InputStream stream = Files.newInputStream(Paths.get(args[1]), StandardOpenOption.READ)) {
                JarInputStream jis = new JarInputStream(stream);
                mainClassName = jis.getManifest().getMainAttributes().getValue("Main-Class");
            }

            if (mainClassName == null) {
                System.err.println("JAR file does not have a Main-Class attribute, is not executable.");
                System.exit(1);
            }
            // Use the parent classloader to forcibly toss out this version of the interop JAR, to avoid confusion
            // later when there are two classloaders in play, BUT, holepunch this specific class and inner classes
            // through so the state linked up from the bootstrap script is still here. In other words, this class is
            // special, so don't give it dependencies outside the JDK.
            ClassLoader thisClassLoader = NodeJS.class.getClassLoader();
            URLClassLoader child = new URLClassLoader(new URL[] {url}, thisClassLoader.getParent()) {
                @Override
                protected Class<?> findClass(String name) throws ClassNotFoundException {
                    if (name.startsWith(NodeJS.class.getName())) {
                        return thisClassLoader.loadClass(name);
                    } else
                        return super.findClass(name);
                }
            };
            Class<?> entryPoint = Class.forName(mainClassName, true, child);
            startJavaThread(entryPoint, Arrays.copyOfRange(args, 2, args.length));
        }
    }

    private static void startJavaThread(Class<?> entryPoint, String[] args) {
        Thread javaThread = new Thread(() -> {
            try {
                Method main = entryPoint.getMethod("main", String[].class);
                assert Modifier.isStatic(main.getModifiers());
                main.invoke(null, new Object[] { args });
                System.exit(0);
            } catch (NoSuchMethodException e) {
                System.err.println("No main method found in " + entryPoint.getName());
            } catch (InvocationTargetException e) {
                e.getCause().printStackTrace();
            } catch (Throwable e) {
                e.printStackTrace();
            }
            System.exit(1);
        }, "Java main thread");
        javaThread.start();
    }

    /**
     * Returns true if executing within the main NodeJS thread. Note: does NOT return true if you are executing on
     * some other worker you created yourself.
     */
    public static boolean isOnMainNodeThread() {
        return Thread.currentThread() == linkage.nodeJSThread;
    }

    /**
     * Throws {@link IllegalStateException} if you aren't on the NodeJS main thread.
     */
    public static void checkOnMainNodeThread() {
        if (!isOnMainNodeThread())
            throw new IllegalStateException("You are not currently on the NodeJS thread and thus cannot access the JavaScript world. Surround your calls to JS with NodeJS.runJS(), NodeJS.runJSAsync() or the Kotlin nodejs{} block.");
    }

    /**
     * This {@link Executor} runs the given commands on the NodeJS event loop thread. The other utility functions
     * on this class are all just utilities that delegate to this executor.
     */
    public static Executor executor = new Executor() {
        @Override
        public void execute(@NotNull Runnable command) {
            if (linkage == null)
                throw new IllegalStateException("This JVM was not started with the nodejvm script.");

            if (Thread.currentThread() == linkage.nodeJSThread)
                command.run();
            else
                linkage.taskQueue.add(command);
        }
    };

    /**
     * Schedules execution of the given {@link Supplier} onto the NodeJS thread, and returns a future that will
     * complete when it's been run.
     */
    public static <T> CompletableFuture<T> runJSAsync(Supplier<T> callable) {
        return CompletableFuture.supplyAsync(callable, executor);
    }

    /**
     * Runs the given {@link Supplier} on the NodeJS thread, blocks until execution has been performed and
     * returns the result that was computed.
     */
    public static <T> T runJS(Supplier<T> callable) {
        try {
            return runJSAsync(callable).get();
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        } catch (ExecutionException e) {
            if (e.getCause() instanceof RuntimeException)
                throw (RuntimeException) e.getCause();
            else
                throw new RuntimeException(e.getCause());
        }
    }

    /**
     * Evaluates the given string and returns a generic {@link Value} object representing the result, that can then
     * be converted into more useful forms. Note that you must be on the NodeJS thread for this to work (see
     * {@link #runJS(Supplier)} for how to do this).
     *
     * @throws IllegalStateException if you're not on the NodeJS thread.
     */
    public static Value eval(@Language("JavaScript") String js) {
        checkOnMainNodeThread();
        return linkage.evalFunction.execute(js);
    }

    /**
     * Returns the {@link Context} for the NodeJS main thread.
     */
    public static Context polyglotContext() {
        return linkage.ctx;
    }

    /**
     * Converts the {@link Value} to a JVM type in the following way:<p>
     *
     * <ol>
     * <li> If the type is an interface not annotated with {@code @FunctionalInterface} then a special proxy is returned that
     *    knows how to map JavaBean style property methods on that interface to JavaScript properties.</li>
     * <li> Otherwise, the {@link Value#as} method is used with the {@link TypeLiteral} so generics are preserved and
     *    the best possible translation occurs.</li>
     * </ol>
     */
    public static <T> T castValue(Value value, TypeLiteral<T> typeLiteral) {
        return NodeJSAPI.castValue(value, typeLiteral);
    }
}

File: src/main/java/net/plan99/nodejs/kotlin/NodeJS.kt
@file:Suppress("UNUSED_VARIABLE")

package net.plan99.nodejs.kotlin

import net.plan99.nodejs.NodeJS
import org.graalvm.polyglot.TypeLiteral
import org.graalvm.polyglot.Value
import org.intellij.lang.annotations.Language
import java.lang.reflect.InvocationHandler
import java.lang.reflect.Method
import java.lang.reflect.Proxy
import kotlin.reflect.KProperty

/**
 * Enters the NodeJS event loop thread and makes the JavaScript API available in scope. You can return JavaScript
 * objects from a nodejs block, however, you can't access them again until you're back inside. Attempting to use
 * them outside a nodejs block will trigger a threading violation exception.
 *
 * Note that this function schedules the lambda onto the NodeJS event loop and waits for it to be executed. If Node
 * is busy, it'll block and wait for it.
 */
fun <T> nodejs(body: NodeJSBlock.() -> T): T = NodeJS.runJS { body(NodeJSBlock()) }

/**
 * Defines various Kotlin helpers to make working with the JavaScript environment more pleasant. Intended to be used
 * inside a [nodejs] block.
 */
class NodeJSBlock internal constructor() {
    /**
     * Converts the [Value] to a JVM type [T] in the following way:
     *
     * 1. If the type is an interface not annotated with `@FunctionalInterface` then a special proxy is returned that
     *    knows how to map JavaBean style property methods on that interface to JavaScript properties.
     * 2. Otherwise, the [Value. as] method is used with a [TypeLiteral] so generics are preserved and the best possible
     *    translation occurs.
     */
    inline fun <reified T> Value.cast(): T = castValue(this, object : TypeLiteral<T>() {})

    companion object {
        /** @see Value.cast */
        @JvmStatic
        fun <T> castValue(value: Value, typeLiteral: TypeLiteral<T>): T {
            val clazz = typeLiteral.rawType
            @Suppress("UNCHECKED_CAST")
            return if (JSTranslationProxyHandler.isTranslateableInterface(clazz))
                Proxy.newProxyInstance(clazz.classLoader, arrayOf(clazz, *clazz.interfaces), JSTranslationProxyHandler(value)) as T
            else
                value.`as`(typeLiteral)
        }
    }

    /** Casts any object to being a JavaScript object. */
    fun Any.asValue(): Value = NodeJS.polyglotContext().asValue(this)

    /**
     * Evaluates the given JavaScript string and casts the result to the desired JVM type. You can request a cast
     * to interfaces that map to JS objects, collections, the Graal/Polyglot [Value] type, boxed primitives and more.
     */
    inline fun <reified T> eval(@Language("JavaScript") javascript: String): T = NodeJS.eval(javascript).cast()

    /**
     * Evaluates the given JavaScript but throws away any result.
     */
    fun run(@Language("JavaScript") javascript: String) {
        NodeJS.eval(javascript)
    }

    /** Allows you to read JS properties of the given [Value] using Kotlin indexing syntax. */
    inline operator fun <reified T> Value.get(key: String): T = getMember(key).cast()

    /** Allows you to read JS properties of the given [Value] using Kotlin indexing syntax. */
    operator fun Value.get(key: String): Value = getMember(key)

    /** Allows you to set JS properties of the given [Value] using Kotlin indexing syntax. */
    operator fun Value.set(key: String, value: Any?) = putMember(key, value)

    private val bindings = NodeJS.polyglotContext().polyglotBindings

    /** Implementation for [bind]. The necessary operator functions are defined as extensions to allow for reified generics. */
    class Binding<T>

    operator fun <T> Binding<T>.setValue(thisRef: Any?, property: KProperty<*>, value: T) {
        // This rather ugly hack is required as we can't just insert the name directly,
        // we have to go via an intermediate 'bindings' map.
        bindings["__nodejvm_transfer"] = value
        NodeJS.eval("${property.name} = Polyglot.import('__nodejvm_transfer');")
        bindings.removeMember("__nodejvm_transfer")
    }

    inline operator fun <reified T> Binding<T>.getValue(thisRef: Any?, property: KProperty<*>): T = eval(property.name)

    inner class Binder<T>(private val default: T? = null) {
        operator fun provideDelegate(thisRef: Any?, prop: KProperty<*>): Binding<T> {
            val b = Binding<T>()
            if (default != null)
                b.setValue(null, prop, default)
            return b
        }
    }

    /**
     * Use this in property delegate syntax to access top level global variables in the NodeJS context. By declaring
     * a variable as `var x: String by bind()` you can read and write the 'x' global variable in JavaScript world.
     */
    fun <T> bind(default: T? = null) = Binder(default)
}

/** Wraps JS objects with some Bean property convenience glue. */
private class JSTranslationProxyHandler(private val value: Value) : InvocationHandler {
    companion object {
        fun isTranslateableInterface(c: Class<*>) =
            c.isInterface && !c.isAnnotationPresent(FunctionalInterface::class.java)
    }

    // This code does a lot of redundant work on every method call and could be optimised with caches.
    init {
        check(nodejs { value.hasMembers() }) { "Cannot translate this value to an interface because it has no members." }
    }

    override fun invoke(proxy: Any, method: Method, args: Array<out Any>?): Any? {
        // Apply Bean-style naming pattern matching.
        val name = method.name
        fun hasPropName(p: Int) = name.length > p && name[p].isUpperCase()
        val getter = name.startsWith("get") && hasPropName(3)
        val setter = name.startsWith("set") && hasPropName(3)
        val izzer = name.startsWith("is") && hasPropName(2)
        val isPropAccess = getter || setter || izzer
        val propName = if (isPropAccess) {
            if (getter || setter) {
                name.drop(3).decapitalize()
            } else {
                check(izzer)
                name.drop(2).decapitalize()
            }
        } else null

        val returnType = method.returnType
        val parameterCount = method.parameterCount

        when {
            izzer -> check(returnType == Boolean::class.java && parameterCount == 0) {
                "Methods starting with 'is' should return boolean and have no parameters."
            }
            getter -> check(parameterCount == 0) { "Methods starting with 'get' should not have any parameters." }
            setter -> check(parameterCount == 1) { "Methods starting with 'set' should have a single parameter." }
        }

        NodeJS.checkOnMainNodeThread()

        return if (propName != null) {
            if (getter || izzer) {
                val member = value.getMember(propName)
                    ?: throw IllegalStateException("No property with name $propName found: [${value.memberKeys}] ")
                member.`as`(returnType)
            } else {
                check(setter)
                value.putMember(propName, args!!.single())
                null
            }
        } else {
            // Otherwise treat it as a method call.
            check(value.canInvokeMember(name)) { "Method $name does not appear to map to an executable member: [${value.memberKeys}]" }
            val result = value.invokeMember(name, *(args ?: emptyArray<Any>()))
            // The result should be thrown out if expecting void, or translated again if the return type is a
            // non-functional interface (functional interfaces are auto-translated by Polyglot already), or
            // otherwise we just rely on the default Polyglot handling which is pretty good most of the time.
            when {
                returnType == Void.TYPE -> null
                isTranslateableInterface(returnType) -> Proxy.newProxyInstance(
                    this.javaClass.classLoader,
                    returnType.interfaces,
                    JSTranslationProxyHandler(result)
                )
                else -> result.`as`(returnType)
            }
        }
    }
}

